test/com/esotericsoftware/kryo/KryoTestCase.java

```diff
@@ -173,6 +175,12 @@ public Input createInput (byte[] buffer) {
 
 	/** @param length Pass Integer.MIN_VALUE to disable checking the length. */
 	public <T> T roundTripWithBufferFactory (int length, T object1, BufferFactory sf) {
+		roundTripWithBufferFactory(length, object1, sf, NoGenericsStrategy.INSTANCE);
+		return roundTripWithBufferFactory(length, object1, sf, new DefaultGenericsStrategy(kryo));
+	}
+
+	/** @param length Pass Integer.MIN_VALUE to disable checking the length. */
+	public <T> T roundTripWithBufferFactory (int length, T object1, BufferFactory sf, GenericsStrategy genericStrategy) {

```

Comment:
It seems you never actually apply the `GenericsStrategy` to the Kryo instance. If you did, I guess that most of the tests would fail because the `length` does not match when `NoGenericsStrategy` is used.

While I like the idea of executing as many tests as possible with the new strategy, it would require two different `length` parameters for every call and is not very intuitive. I think it would be easier for now if you just craft a single test-case that calls all or most of the methods on `NoGenericsStrategy` and put it in `GenericsTest`.

@magro: WDYT?

Reply:
Oh, you're right, my fault. Hmm... So, what we could do is using both strategies for all non-generic test cases. The generic test cases can specify the strategy to use.
I think, currently there is no test case faiing with the current default generics strategy, right? So the previous test cases could specify this strategy directly.

Reply:
> So, what we could do is using both strategies for all non-generic test cases. The generic test cases can specify the strategy to use.

It would probably be enough to create an additional `roundTrip` method with two length parameters and use that explicitly for all generics tests. I looked through the previous implementation of `setOptimizedGenerics` and I think it was only used for generics tests as well. (https://github.com/EsotericSoftware/kryo/commit/167ef87457ea8fa624e87312ca7680ff97a1bb92)

> I think, currently there is no test case faiing with the current default generics strategy, right? 

Correct. Some additional tests were merged into master yesterday, so you might want to pull the changes into your branch.

Reply:
But let's wait for feedback from @magro as well.

Reply:
I've pulled from master and implemented the changes, so it should work well now.
I am still open for feedback from you and magro :)

Reply:
Great thanks! 

Can you please do me a favor and revert all changes not directly related to your PR? It makes it so much harder to review with all the added `@Override` etc. It definitely makes sense to add those but we should do this in a separate PR that doesn't contain any logic changes. Also your import order doesn't seem to match the project import order. Please check this at well.

Reply:
https://github.com/EsotericSoftware/kryo/blob/master/eclipse/.settings/org.eclipse.jdt.ui.prefs references a formatter profile libgdx. You may want to change that.
You can also enable project-specific code formatter options so that user do not have to do that manually. You can also enable a project-specific code-clean up, which can, among other things, automatically introduce @Overrides to missing methods.
Via Save actions (also in the project settings) these actions can be performed on save automatically, leading to a coherent source code (at least for eclipse users).
You did not specify a particular imports-order. These can be specified in the "Organize Imports" settings page in Eclipse and are saved with an ".importorder" file extension by default. I do not want to edit the imports by hand.

If you are keen on enforcing specific formatting rules, you may consider implementing an approach similar to Soot (https://github.com/Sable/soot), where pull requests are checked for style violations upon merge by an automatic process. I'd highly recommend that and supplying as much automation as possible. Nobody wants to change things like import order by hand...
See https://github.com/Sable/soot/tree/master/codingstyle for a reference on what files you'd needed to supply. Apart from that, as I said, I think preconfiguring the project and committing these rules as well into the repository seems like the easiest way to ensure everyone's on the same page.

I've imported the code-format.xml in Eclipse and ran it over the whole code and commited the result. Some files where not following the code formatting prior either: https://github.com/EsotericSoftware/kryo/pull/683/commits/ef34f1faaf604a8b085915e98c44e47abf795c3d


I'll remove the @Overrides visible in the diff by hand, but honestly, I do not think it is a particular good idea, especially when you plan to introduce them later anyway.

Reply:
I fully agree with everything you suggested! I normally use Sonarqube in my projects to enforce code style and quality rules. 

The issue is this: I'm not a developer of Kryo. I'm just a user who likes the project, has been using Kryo 4 for years and wants to see a stable final release of Kryo 5. @magro is the only one currently available that has rights to merge PRs and perform releases. So we have to make it as easy for him as possible to review the PRs and decide whether they can be merged or not. This is why I'm trying to reduce the number of changes in your PR to the absolute minimum necessary to solve the issue at hand.

It would be great if you could help with code style settings and code quality automation after this is merged!

Reply:
@magro: Are you using Eclipse for working on Kryo? We need a working `.importorder` file to avoid imports getting shuffled. 

I'm using IntelliJ with Eclipse formatter and it seems to work fine.

Reply:
The current import order seems to be something like this:

```
static all other, 
blank, 
com.*, 
blank, 
java.*, 
blank, 
javax.*, 
blank, 
org.*, 
blank, 
all other imports
```

Reply:
Nevermind! I just handcrafted the `.importorder` file: https://github.com/EsotericSoftware/kryo/pull/726/commits/52f272ed8a89d5bbc5c9b01fe2c7eab7501c735f

 I'm not sure how `*` imports are configured in Eclipse and if this setting is included in the current `codeformat.xml`. Static imports always use `*` in the current codebase.

@MarcMil please import it and apply it to your PR.

Reply:
Great! I'm using intellij, Nate is using Eclipse AFAIK.